File: _deprecated/api-server/src/routes/games.ts
import { Router } from "express";
import { GameModel, UserModel, type GameAttrs } from "@workspace/db";
import { logger } from "../lib/logger";

const router = Router();

const CATEGORY_LABELS: Record<string, { label: string; emoji: string }> = {
  matematica: { label: "Matematică", emoji: "🔢" },
  litere: { label: "Litere & Alfabet", emoji: "📚" },
  culori: { label: "Culori & Forme", emoji: "🎨" },
  muzica: { label: "Muzică & Ritm", emoji: "🎵" },
  natura: { label: "Natură & Animale", emoji: "🌿" },
  logica: { label: "Logică & Puzzle", emoji: "🧩" },
  memorie: { label: "Memorie & Concentrare", emoji: "🧠" },
  creativitate: { label: "Creativitate", emoji: "✏️" },
};

// The wire format calls this field `isNew`; it's stored as `isNewGame` in
// Mongo to avoid colliding with Mongoose's built-in Document#isNew. See
// lib/db/src/schema/games.ts for the full explanation.
function serializeGame(g: GameAttrs) {
  return {
    id: g.id,
    title: g.title,
    description: g.description,
    category: g.category,
    ageMin: g.ageMin,
    ageMax: g.ageMax,
    imageUrl: g.imageUrl,
    gameUrl: g.gameUrl ?? null,
    isNew: g.isNewGame,
    isFeatured: g.isFeatured,
    requiresSubscription: g.requiresSubscription,
    createdAt: new Date(g.createdAt).toISOString(),
  };
}

router.get("/games", async (req, res) => {
  try {
    const { category, ageMin, ageMax, featured } = req.query;
    const filter: Record<string, unknown> = {};

    if (category) filter.category = String(category);
    if (featured === "true") filter.isFeatured = true;

    // Mirrors the original Postgres query: a game matches when its own
    // [ageMin, ageMax] range overlaps the requested [ageMin, ageMax] range.
    if (ageMin) filter.ageMax = { $gte: Number(ageMin) };
    if (ageMax) filter.ageMin = { $lte: Number(ageMax) };

    const games = await GameModel.find(filter).sort({ createdAt: 1 }).lean<GameAttrs[]>();
    res.json(games.map(serializeGame));
  } catch (err) {
    logger.error({ err }, "Failed to list games");
    res.status(500).json({ error: "Failed to list games" });
  }
});

router.get("/games/featured", async (req, res) => {
  try {
    const games = await GameModel.find({ isFeatured: true })
      .sort({ createdAt: 1 })
      .lean<GameAttrs[]>();
    res.json(games.map(serializeGame));
  } catch (err) {
    logger.error({ err }, "Failed to list featured games");
    res.status(500).json({ error: "Failed to list featured games" });
  }
});

router.get("/games/categories", async (req, res) => {
  try {
    const rows = await GameModel.aggregate<{ _id: string; count: number }>([
      { $group: { _id: "$category", count: { $sum: 1 } } },
      { $sort: { count: -1 } },
    ]);
    const categories = rows.map((r) => ({
      name: r._id,
      label: CATEGORY_LABELS[r._id]?.label ?? r._id,
      count: r.count,
      emoji: CATEGORY_LABELS[r._id]?.emoji ?? "🎮",
    }));
    res.json(categories);
  } catch (err) {
    logger.error({ err }, "Failed to list categories");
    res.status(500).json({ error: "Failed to list categories" });
  }
});

router.get("/games/new", async (req, res) => {
  try {
    const games = await GameModel.find({ isNewGame: true })
      .sort({ createdAt: 1 })
      .lean<GameAttrs[]>();
    res.json(games.map(serializeGame));
  } catch (err) {
    logger.error({ err }, "Failed to list new games");
    res.status(500).json({ error: "Failed to list new games" });
  }
});

router.get("/games/:id", async (req, res) => {
  try {
    const id = Number(req.params.id);
    if (!Number.isFinite(id)) return res.status(404).json({ error: "Game not found" });

    const game = await GameModel.findOne({ id }).lean<GameAttrs>();
    if (!game) return res.status(404).json({ error: "Game not found" });
    res.json(serializeGame(game));
  } catch (err) {
    logger.error({ err }, "Failed to get game");
    res.status(500).json({ error: "Failed to get game" });
  }
});

router.get("/stats/overview", async (req, res) => {
  try {
    const thirtyDaysAgo = new Date(Date.now() - 30 * 24 * 60 * 60 * 1000);

    const [totalGames, distinctCategories, totalUsers, newGamesThisMonth] = await Promise.all([
      GameModel.countDocuments(),
      GameModel.distinct("category"),
      UserModel.countDocuments(),
      GameModel.countDocuments({ createdAt: { $gte: thirtyDaysAgo } }),
    ]);

    res.json({
      totalGames,
      totalCategories: distinctCategories.length,
      totalUsers,
      newGamesThisMonth,
    });
  } catch (err) {
    logger.error({ err }, "Failed to get stats");
    res.status(500).json({ error: "Failed to get stats" });
  }
});

export default router;

File: artifacts/api-server-py/app/routers/games.py
"""Port of artifacts/api-server/src/routes/games.ts — same endpoints, same
query params (camelCase, to match the existing OpenAPI contract / generated
frontend client unchanged), same response shape.
"""

from datetime import datetime, timedelta, timezone

from fastapi import APIRouter, HTTPException, Query

from app.logger import log_error
from app.models.game import Game
from app.models.user import User

router = APIRouter()

CATEGORY_LABELS = {
    "matematica": {"label": "Matematică", "emoji": "🔢"},
    "litere": {"label": "Litere & Alfabet", "emoji": "📚"},
    "culori": {"label": "Culori & Forme", "emoji": "🎨"},
    "muzica": {"label": "Muzică & Ritm", "emoji": "🎵"},
    "natura": {"label": "Natură & Animale", "emoji": "🌿"},
    "logica": {"label": "Logică & Puzzle", "emoji": "🧩"},
    "memorie": {"label": "Memorie & Concentrare", "emoji": "🧠"},
    "creativitate": {"label": "Creativitate", "emoji": "✏️"},
    "labirint":     {"label": "Labirint",    "emoji": "🌀"},
    "cuvinte":      {"label": "Caută Cuvinte", "emoji": "🔍"},
}


def serialize_game(g: Game) -> dict:
    # Wire format calls this field `isNew`; stored as `is_new_game` /
    # Mongo `isNewGame` — see app/models/game.py for why.
    return {
        "id": g.numeric_id,
        "title": g.title,
        "description": g.description,
        "category": g.category,
        "ageMin": g.age_min,
        "ageMax": g.age_max,
        "imageUrl": g.image_url,
        "gameUrl": g.game_url,
        "isNew": g.is_new_game,
        "isFeatured": g.is_featured,
        "requiresSubscription": g.requires_subscription,
        "createdAt": g.created_at.isoformat(),
    }


@router.get("/games")
async def list_games(
    category: str | None = None,
    age_min: int | None = Query(None, alias="ageMin"),
    age_max: int | None = Query(None, alias="ageMax"),
    featured: str | None = None,
):
    try:
        conditions = []
        if category:
            conditions.append(Game.category == category)
        if featured == "true":
            conditions.append(Game.is_featured == True)  # noqa: E712
        # Mirrors the original query: a game matches when its own
        # [ageMin, ageMax] range overlaps the requested [ageMin, ageMax].
        if age_min is not None:
            conditions.append(Game.age_max >= age_min)
        if age_max is not None:
            conditions.append(Game.age_min <= age_max)

        query = Game.find(*conditions) if conditions else Game.find_all()
        games = await query.sort(Game.created_at).to_list()
        return [serialize_game(g) for g in games]
    except Exception as err:  # noqa: BLE001
        log_error("Failed to list games", err=str(err))
        raise HTTPException(status_code=500, detail="Failed to list games")


@router.get("/games/featured")
async def list_featured_games():
    try:
        games = await Game.find(Game.is_featured == True).sort(Game.created_at).to_list()  # noqa: E712
        return [serialize_game(g) for g in games]
    except Exception as err:  # noqa: BLE001
        log_error("Failed to list featured games", err=str(err))
        raise HTTPException(status_code=500, detail="Failed to list featured games")


@router.get("/games/categories")
async def list_categories():
    try:
        raw = Game.get_pymongo_collection()
        rows = await raw.aggregate(
            [
                {"$group": {"_id": "$category", "count": {"$sum": 1}}},
                {"$sort": {"count": -1}},
            ]
        ).to_list(length=None)

        return [
            {
                "name": row["_id"],
                "label": CATEGORY_LABELS.get(row["_id"], {}).get("label", row["_id"]),
                "count": row["count"],
                "emoji": CATEGORY_LABELS.get(row["_id"], {}).get("emoji", "🎮"),
            }
            for row in rows
        ]
    except Exception as err:  # noqa: BLE001
        log_error("Failed to list categories", err=str(err))
        raise HTTPException(status_code=500, detail="Failed to list categories")


@router.get("/games/new")
async def list_new_games():
    try:
        games = await Game.find(Game.is_new_game == True).sort(Game.created_at).to_list()  # noqa: E712
        return [serialize_game(g) for g in games]
    except Exception as err:  # noqa: BLE001
        log_error("Failed to list new games", err=str(err))
        raise HTTPException(status_code=500, detail="Failed to list new games")


@router.get("/games/{game_id}")
async def get_game(game_id: str):
    try:
        try:
            numeric_id = int(game_id)
        except ValueError:
            raise HTTPException(status_code=404, detail="Game not found")

        game = await Game.find_one(Game.numeric_id == numeric_id)
        if not game:
            raise HTTPException(status_code=404, detail="Game not found")
        return serialize_game(game)
    except HTTPException:
        raise
    except Exception as err:  # noqa: BLE001
        log_error("Failed to get game", err=str(err))
        raise HTTPException(status_code=500, detail="Failed to get game")


@router.get("/stats/overview")
async def stats_overview():
    try:
        thirty_days_ago = datetime.now(timezone.utc) - timedelta(days=30)

        total_games = await Game.find_all().count()
        total_users = await User.find_all().count()
        new_games_this_month = await Game.find(Game.created_at >= thirty_days_ago).count()
        raw = Game.get_pymongo_collection()
        distinct_categories = await raw.distinct("category")

        return {
            "totalGames": total_games,
            "totalCategories": len(distinct_categories),
            "totalUsers": total_users,
            "newGamesThisMonth": new_games_this_month,
        }
    except Exception as err:  # noqa: BLE001
        log_error("Failed to get stats", err=str(err))
        raise HTTPException(status_code=500, detail="Failed to get stats")
